noir-projects/noir-contracts/contracts/token_with_refunds_contract/src/main.nr

```diff
@@ -165,17 +441,17 @@ contract TokenWithRefunds {
         // the authwit flow here and check that the user really permitted fee_payer to set up a refund on their behalf.
         assert_current_call_valid_authwit(&mut context, user);
 
-        // 2. Get all the relevant user keys
+        // 2. Get all the relevant keys
         let header = context.get_header();
+        let fee_payer_npk_m_hash = header.get_npk_m_hash(&mut context, fee_payer);
         let user_npk_m_hash = header.get_npk_m_hash(&mut context, user);
         let user_ovpk = header.get_ovpk_m(&mut context, user);
         let user_ivpk = header.get_ivpk_m(&mut context, user);
 
         // 3. Deduct the funded amount from the user's balance - this is a maximum fee a user is willing to pay
         // (called fee limit in aztec spec). The difference between fee limit and the actual tx fee will be refunded 
         // to the user in the `complete_refund(...)` function.
-        // TODO(#7324), TODO(#7323): using npk_m_hash here is vulnerable in 2 ways described in the linked issues.
-        storage.balances.sub(user_npk_m_hash, U128::from_integer(funded_amount)).emit(encode_and_encrypt_note_with_keys(&mut context, user_ovpk, user_ivpk, user));
+        storage.balances.sub(user, U128::from_integer(funded_amount)).emit(encode_and_encrypt_note_with_keys(&mut context, user_ovpk, user_ivpk, user));
```

Comment:
Now we store the notes under user address here and not under npk_m_hash which is another piece of key rotation fix.